### ISSUE TITLE
dart2js incorrectly fuses statements together 

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#54823]
The code below is miscompiled by dart2js, which decides to emit `this.buf += this.recurse$1(i - 1);` in `recurse` rather than storing result of `recurse` in a temporary variable and then doing `this.buf += t`. Such fusion is incorrect because `recurse` has side-effect on `buf`:

```dart
class V {
  String buf = "";

  void write(String value) {
    buf += value;
  }

  String recurse(int i) {
    if (i > 0) {
      write(recurse(i - 1));
      return '[$i]';
    } else {
      return '[0]';
    }
  }
}

void main() {
  final v = V();
  v.recurse(2);
  print(v.buf);
}
```

```console
╭─~/s/d/sdk ⟨main ✔⟩
╰─» dart test.dart
[0][1]
╭─~/s/d/sdk ⟨main ✔⟩
╰─» dart compile js -O4 --no-minify test.dart
Compiled 10,080,692 input bytes (5,082,436 characters source) to 12,702 characters JavaScript in 0.30 seconds using 180.484 MB of memory
╭─~/s/d/sdk ⟨main ✔⟩
╰─» node out.js
[1]
```

(This issue was originally reported by @ykmnkmi on Discord)

**Comment by ykmnkmi** [comment]
@mraleph, thanks. For context, [Here](https://github.com/ykmnkmi/jinja.dart/blob/main/test/statements/for_test.dart#L164) are the recursive tests that fails on the web. This was fixed by adding the `@noInline` annotation [here](https://github.com/ykmnkmi/jinja.dart/blob/main/lib/src/renderer.dart#L122).

**Comment by rakudrama** [comment]
Fixed by d8456b7cd8a30b5790c48b90502026c3fe679d01
Thanks for the great repo.
